File: bot-socket/app/agent/streaming.py
"""
AI SDK v4 line protocol formatters.

This module produces the exact same SSE line format used by Metabase's
agent/streaming.clj:

  0:"text delta"                         — text part
  9:{toolCallId, toolName, args}         — tool call part
  a:{toolCallId, result}                 — tool result part
  2:[{type, data}]                       — structured data part (state, navigate, etc.)
  3:"error message"                      — error part
  d:{finishReason, usage}                — finish part

The frontend lib/streaming.ts parses these exact prefixes.
"""

from __future__ import annotations

import json


def _j(obj: object) -> str:
    return json.dumps(obj, ensure_ascii=False, separators=(",", ":"))


def text_part(text: str) -> str:
    """Stream a text delta to the client (prefix 0)."""
    return f"0:{_j(text)}\n"


def tool_call_part(tool_call_id: str, tool_name: str, args: dict) -> str:
    """Notify the client that the LLM is invoking a tool (prefix 9)."""
    return f"9:{_j({'toolCallId': tool_call_id, 'toolName': tool_name, 'args': args})}\n"


def tool_result_part(tool_call_id: str, result: str) -> str:
    """Send the result of a tool back to the client (prefix a)."""
    return f"a:{_j({'toolCallId': tool_call_id, 'result': result})}\n"


def data_part(data_type: str, data: object) -> str:
    """Send structured data (state, navigate_to, etc.) to the client (prefix 2)."""
    return f"2:{_j([{'type': data_type, 'data': data}])}\n"


def error_part(message: str) -> str:
    """Send an error to the client (prefix 3)."""
    return f"3:{_j(message)}\n"


def finish_part(
    finish_reason: str = "stop",
    prompt_tokens: int = 0,
    completion_tokens: int = 0,
) -> str:
    """Signal stream completion with usage stats (prefix d)."""
    return f"d:{_j({'finishReason': finish_reason, 'usage': {'promptTokens': prompt_tokens, 'completionTokens': completion_tokens}})}\n"

File: bot-ui/src/lib/streaming.ts
/**
 * AI SDK v4 line-protocol stream parser.
 *
 * Exact port of Metabase's frontend/src/metabase/api/ai-streaming/process-stream.ts
 *
 * Line format (each SSE data line is one of):
 *   0:"text"                    — text delta
 *   9:{toolCallId,toolName,args} — tool call
 *   a:{toolCallId,result}        — tool result
 *   2:[{type,data}]              — structured data part (state, navigate, etc.)
 *   3:"error"                   — error
 *   d:{finishReason,usage}       — finish
 */

import type { StreamCallbacks } from "@/store/types";

/**
 * Consume a streaming fetch Response and dispatch typed callbacks for
 * each AI SDK v4 line-protocol part.
 *
 * @param response  A fetch Response with a readable body (text/event-stream).
 * @param callbacks Object of typed handlers (see StreamCallbacks).
 */
export async function processStream(
  response: Response,
  callbacks: StreamCallbacks
): Promise<void> {
  if (!response.body) {
    callbacks.onErrorPart("Response has no body");
    return;
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });

      // Split on newlines — each complete line is one protocol part
      const lines = buffer.split("\n");
      // Keep the last (possibly incomplete) chunk in the buffer
      buffer = lines.pop() ?? "";

      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;

        const colonIdx = trimmed.indexOf(":");
        if (colonIdx === -1) continue;

        const prefix = trimmed.slice(0, colonIdx);
        const payload = trimmed.slice(colonIdx + 1);

        try {
          switch (prefix) {
            case "0": {
              // Text delta
              const text = JSON.parse(payload) as string;
              callbacks.onTextPart(text);
              break;
            }

            case "9": {
              // Tool call notification
              const { toolCallId, toolName, args } = JSON.parse(payload) as {
                toolCallId: string;
                toolName: string;
                args: Record<string, unknown>;
              };
              callbacks.onToolCallPart(toolCallId, toolName, args);
              break;
            }

            case "a": {
              // Tool result
              const { toolCallId, result } = JSON.parse(payload) as {
                toolCallId: string;
                result: string;
              };
              callbacks.onToolResultPart(toolCallId, result);
              break;
            }

            case "2": {
              // Structured data array — may contain multiple items
              const dataArray = JSON.parse(payload) as Array<{
                type: string;
                data: unknown;
              }>;
              for (const item of dataArray) {
                callbacks.onDataPart(item.type, item.data);
              }
              break;
            }

            case "3": {
              // Error
              const errorMsg = JSON.parse(payload) as string;
              callbacks.onErrorPart(errorMsg);
              break;
            }

            case "d": {
              // Finish — includes usage stats
              const { finishReason, usage } = JSON.parse(payload) as {
                finishReason: string;
                usage: { promptTokens: number; completionTokens: number };
              };
              callbacks.onFinishPart(finishReason, usage);
              break;
            }

            default:
              // Unknown prefix — ignore (forward compatibility)
              break;
          }
        } catch (parseError) {
          console.warn("[processStream] failed to parse line:", trimmed, parseError);
        }
      }
    }
  } finally {
    reader.releaseLock();
  }
}
